scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -788,16 +790,112 @@ void BPF_STRUCT_OPS(layered_enqueue, struct task_struct *p, u64 enq_flags)
 			return;
 	}
 
-	bpf_for(idx, 0, nr_possible_cpus) {
-		s32 cand = (preempt_cursor + idx) % nr_possible_cpus;
+	if (!disable_topology) {
+		if (!(cachec = lookup_cache_ctx(cctx->cache_idx)) ||
+		    !(nodec = lookup_node_ctx(cctx->node_idx)))
+			return;
 
-		if (try_preempt(cand, p, cctx, tctx, layer, false)) {
-			/*
-			 * Round-robining doesn't have to be strict. Let's
-			 * not bother with atomic ops on $preempt_cursor.
-			 */
-			preempt_cursor = (cand + 1) % nr_possible_cpus;
+		struct bpf_cpumask *attempted = bpf_cpumask_create();
+		if (!attempted)
 			return;
+
+		struct bpf_cpumask *topo_cpus = bpf_cpumask_create();
+		if (!topo_cpus) {
+			bpf_cpumask_release(attempted);
+			return;
+		}
+
+		if (!cachec->cpumask) {
+			bpf_cpumask_release(attempted);
+			bpf_cpumask_release(topo_cpus);
+			return;
+		}
+
+		bpf_cpumask_copy(topo_cpus, (const struct cpumask*)cachec->cpumask);
+
+		/*
+		 * First try preempting in the local LLC
+		 */
+		bpf_for(idx, 0, cachec->nr_cpus) {
+			s32 preempt_cpu = bpf_cpumask_any_distribute((const struct cpumask*)topo_cpus);
+			if (preempt_cpu > cachec->nr_cpus)
+				break;
+
+			if (try_preempt(preempt_cpu, p, cctx, tctx, layer, false)) {
+				bpf_cpumask_release(attempted);
+				bpf_cpumask_release(topo_cpus);
+				return;
+			}
+			bpf_cpumask_clear_cpu(preempt_cpu, topo_cpus);
+			bpf_cpumask_set_cpu(preempt_cpu, attempted);
+		}
+
+		/*
+		 * Next try node local LLC
+		 */
+		if (!nodec->cpumask) {
+			bpf_cpumask_release(attempted);
+			bpf_cpumask_release(topo_cpus);
+			return;
+		}
+		bpf_cpumask_copy(topo_cpus, (const struct cpumask*)nodec->cpumask);
+		bpf_cpumask_xor(topo_cpus,
+				(const struct cpumask*)attempted,
+				(const struct cpumask*)topo_cpus);
+
+		bpf_for(idx, 0, nodec->nr_cpus) {
+			s32 preempt_cpu = bpf_cpumask_any_distribute((const struct cpumask*)topo_cpus);
+			if (try_preempt(preempt_cpu, p, cctx, tctx, layer, false)) {
+				bpf_cpumask_release(attempted);
+				bpf_cpumask_release(topo_cpus);
+				return;
+			}
+			bpf_cpumask_clear_cpu(preempt_cpu, topo_cpus);
+			bpf_cpumask_set_cpu(preempt_cpu, attempted);
+			if (bpf_cpumask_empty((const struct cpumask*)topo_cpus))
+				break;
+		}
+
+		/*
+		 * Finally try across nodes
+		 */
```

Comment:
It may not even be helpful to preempt across node boundaries. Might be worthwhile to make this configurable and see whether this helps or not.